### PR TITLE
I18n french

### PR DESCRIPTION
**Pull request by ludoo0d0a**
here are french translations
+ general settings lang toggle button
+ upgrade libs

### CODE REVIEW

File: assets/i18n/de.json [hunk 1]
```diff
@@ -1,2931 +1,2807 @@
-{
-  "abaddon_title" : "Abaddon Code",
```
Comment (S-Man42) [inline assets/i18n/de.json:2]: Is there a reason why you completely replaced the file? Maybe a whitespace issue or something?

File: assets/i18n/de.json [hunk 2]
```diff
@@ -1,2931 +1,2807 @@
...
+  "formulasolver_groups_exportgroup": "Formelgruppe exportieren",
+  "formulasolver_groups_editgroup": "Formelgruppe bearbeiten",
+  "formulasolver_groups_removegroup": "Formelgruppe löschen",
+  "formulasolver_groups_imported": "Aus Zwischenablage importiert.",
+  "formulasolver_groups_importfromclipboard": "Import aus Zwischenablage",
+  "formulasolver_groups_importerror": "Keine validen Daten gefunden.",
+  "formulasolver_formulas_values": "Werte",
+  "formulasolver_formulas_newformula": "Formel hinzufügen",
+  "formulasolver_formulas_newformula_hint": "Formel eingeben",
+
```
Comment (S-Man42) [inline assets/i18n/de.json:2024]: "Sprache" ;)

File: lib/widgets/main_menu/general_settings.dart [hunk 3]
```diff
@@ -1,35 +1,42 @@
...
 class GeneralSettingsState extends State<GeneralSettings> {
-
   @override
   Widget build(BuildContext context) {
+    var appLanguage = Provider.of<AppLanguage>(context);
+    //var currentLocale = await appLanguage.fetchLocale();
+
     return Column(
       children: <Widget>[
-        GCWTextDivider(
```
Comment (S-Man42) [inline lib/widgets/main_menu/general_settings.dart:27]: I know, code formatting is something, where everyone has a different opinion and it's no fun to talk about it ;) However, it is hard to identify real changes in the code, when there's a done a lot of formatting. Would it be possible if you try to keep the original formatting if possible? That makes the reviews much more efficient. Or what do you mean? :)

File: lib/widgets/main_menu/general_settings.dart [hunk 4]
```diff
@@ -140,30 +139,56 @@ class GeneralSettingsState extends State<GeneralSettings> {
...
-                  });
-                }
-              ),
-            )
+                child: FutureBuilder<Locale>(
+                    future: appLanguage.fetchLocale(),
+                    builder: (BuildContext context, AsyncSnapshot<Locale> snapshot) {
+                      if (!snapshot.hasData) {
+                        // while data is loading:
+                        return Center(
```
Comment (S-Man42) [inline lib/widgets/main_menu/general_settings.dart:166]: I tried, but I never saw a progress indicator here, unfortunately. However, we recently added a styled progress indicator which fits the design. Would it be possible to add it here? May you could ask @Mike-3 who build this:
https://github.com/S-Man42/GCWizard/blob/1.5.0/lib/widgets/common/gcw_async_executer.dart#L99

Maybe we can extract that Indicator from the AsyncExecuter widget to use it more commonly.

File: lib/widgets/main_menu/general_settings.dart [hunk 5]
```diff
@@ -140,30 +139,56 @@ class GeneralSettingsState extends State<GeneralSettings> {
...
+                        return GCWStatefulDropDownButton(
+                            items: supportedLocales.map((locale) {
+                              return GCWDropDownMenuItem(
+                                value: locale.languageCode ?? "en",
+                                child: i18n(context, locale.languageCode) ??  locale.languageCode,
+                              );
+                            }).toList(),
+                            value: currentLocale.languageCode ?? "en",
+                            onChanged: (newValue) {
+                              //appLanguage.changeLocale(newValue);
```
Comment (S-Man42) [inline lib/widgets/main_menu/general_settings.dart:182]: Please remove unused comments :)

File: lib/widgets/main_menu/general_settings.dart [hunk 6]
```diff
@@ -140,30 +139,56 @@ class GeneralSettingsState extends State<GeneralSettings> {
             });
           },
         ),
-        GCWTextDivider(
-          text: i18n(context, 'settings_general_i18n_title')
-        ),
+        GCWTextDivider(text: i18n(context, 'settings_general_i18n_title')),
         Row(
           children: [
+            Expanded(child: GCWText(text: i18n(context, 'settings_general_i18n_defaultlengthunit'))),
```
Comment (S-Man42) [inline lib/widgets/main_menu/general_settings.dart:145]: Maybe we should change the settings order:
1. Internationalization to the top
2. Within internationalization "languages" first to make this the most important setting and don't make the users scroll for searching the language. It's worth it.

**Review by S-Man42 — COMMENTED** [review]
Great work so far! Well done! Thank you so much. The fallback works fine!

Some minor issues and questions, I believe, we can clarify them.

However, one thing I am missing: 
Currently the app reads the internal device locale and automatically chooses the fitting one if available. So, in Germany, a user gets the German version automatically.

However, in my emulator this didn't work anymore with your code. I always got the English version and had to select FR or DE manually. I fear this is something which many users could keep away from this app: They download it, see: "Uuh, English, I don't understand anything" and close the app (most users wouldn't search in an English setting menu for setting the language... We have many many technophobe users unfortunately)

It would be great to keep the language auto detection. So, as long as the user didn't set the language manually, it gets:
1. The device language
2. If not available, EN

File: assets/i18n/de.json (hunk 1 above)
Comment (ludoo0d0a) [reply assets/i18n/de.json:2]: no, it's a mistake

File: assets/i18n/de.json (hunk 2 above)
Comment (ludoo0d0a) [reply assets/i18n/de.json:2024]: these items were missing, isn't it ? 
i should remove or not ? not clear ;)

File: lib/widgets/main_menu/general_settings.dart (hunk 3 above)
Comment (ludoo0d0a) [reply lib/widgets/main_menu/general_settings.dart:27]: is there a linter ?

File: lib/widgets/main_menu/general_settings.dart (hunk 4 above)
Comment (ludoo0d0a) [reply lib/widgets/main_menu/general_settings.dart:166]: loading is fast, should never happen in fact. may be during <10ms.
but we need to put something instead during loading
or not...
I choose to use a loader but not required here
what do you think ?

File: lib/widgets/main_menu/general_settings.dart (hunk 6 above)
Comment (ludoo0d0a) [reply lib/widgets/main_menu/general_settings.dart:145]: agree
didn't want to be too rude for a first version ;)

File: assets/i18n/de.json (hunk 1 above)
Comment (ludoo0d0a) [reply assets/i18n/de.json:2]: maybe a space issue here only, line is not removed

File: lib/widgets/main_menu/general_settings.dart (hunk 3 above)
Comment (ludoo0d0a) [reply lib/widgets/main_menu/general_settings.dart:27]: dartfmt ? with specific rules ? or not ?

**Comment by ludoo0d0a** [comment]
> Currently the app reads the internal device locale and automatically chooses the fitting one if available. So, in Germany, a user gets the German version automatically.

Tested on your current v1.5. 
I set french on my emulator, clear cache, -> en is displayed.
+ I don't see where in the code the system locale is loaded

but I agree on the feature.

File: lib/widgets/main_menu/general_settings.dart (hunk 3 above)
Comment (ludoo0d0a) [reply lib/widgets/main_menu/general_settings.dart:27]: I lint with dartfmt, but it seems you didn't use it before, aren't you ?

Comment (S-Man42) [reply lib/widgets/main_menu/general_settings.dart:27]: OK, i tried the command "dart format lib". That's quite useful I guess. I will commit a formatting commit after the current PRs... So you can leave this as it is :)

File: assets/i18n/de.json (hunk 2 above)
Comment (S-Man42) [reply assets/i18n/de.json:2024]: No I meant, it is "Sprache", not "Sprächen". You should correct the string :)

I just saw another issue, which might explain the complete file change here: You removed all empty lines. We found them extremely useful for more readability. So every tool could be found in its own visual block. Furthermore I used the line counts in en and de to compare if a translation is missing. That is not possible anymore if in one file the empty lines exists and in the other not ;)

Is it possible to restore the empty lines? (as well in your fr.json file).

File: lib/widgets/main_menu/general_settings.dart (hunk 4 above)
Comment (S-Man42) [reply lib/widgets/main_menu/general_settings.dart]: In my personal opinion, I believe it is not necessary to put this into a FutureBuilder. I believe the change of languages is fast enough to simply use `AppBuilder.of(context).rebuild();` as it is done, e.g., at the ThemeChange above.

File: assets/i18n/de.json (hunk 2 above)
Comment (ludoo0d0a) [reply assets/i18n/de.json:2024]: I can restore 'de', but 'fr' has been exported from crowdin
You cannot automate i18n with blank lines.
if you diff, you can use 'ignore blank lines' to have a human readable display.

Comment (ludoo0d0a) [reply assets/i18n/de.json]: I revert de.json from my 1.5.0 version, add just my unique text for settings_general_i18n_language = Sprache

**Comment by ludoo0d0a** [comment]
> > Currently the app reads the internal device locale and automatically chooses the fitting one if available. So, in Germany, a user gets the German version automatically.
> 
> Tested on your current v1.5.
> I set french on my emulator, clear cache, -> en is displayed.
> 
> * I don't see where in the code the system locale is loaded
> 
> but I agree on the feature.

I use platform locale on first startup, as requested.

**Comment by S-Man42** [comment]
The de.json file is still completely changed for some reasons... As I see, it only should add one line.

Furthermore, there are some conflicts which need to be resolved. 

Then I will enjoy to do the merge :)

**Comment by ludoo0d0a** [comment]
> The de.json file is still completely changed for some reasons... As I see, it only should add one line.
> 
> Furthermore, there are some conflicts which need to be resolved.
> 
> Then I will enjoy to do the merge :)

Sorry, I just wanted to add the single line for "settings_general_i18n_language"
I changed nothing else.

File: assets/i18n/de.json (hunk 2 above)
Comment (ludoo0d0a) [reply assets/i18n/de.json]: fixed, CRLF everybody

**Comment by ludoo0d0a** [comment]
@Mike-3 what do you think about these libs upgrade ?
using the latest ? or your version ?

**Review by ludoo0d0a — COMMENTED** [review]
Reviewed all 9 files for french translation and lang toggle in settings

**Comment by Mike-3** [comment]
So I have no problem with that.
In the last open PR there are no changes to the language file.
French can be taken over by me.

**Pull request merged**